sklearn/datasets/_arff_parser.py

```diff
@@ -265,4 +301,174 @@ def _liac_arff_parser(
             y = None
 
-    return X, y, frame, nominal_attributes
+    if output_arrays_type == "pandas":
+        return X, y, frame, None
+    return X, y, None, categories
+
+
+def _pandas_arff_parser(
+    gzip_file,
+    output_type,
+    openml_columns_info,
+    feature_names_to_select,
+    target_names_to_select,
+):
+    """ARFF parser using `pandas.read_csv`.
+
+    This parser uses the metadata fetched directly from OpenML and skips the metadata
+    headers of ARFF file itself. The data is loaded as a CSV file.
+
+    Parameters
+    ----------
+    gzip_file : GzipFile instance
+        The GZip compressed file with the ARFF formatted payload.
+
+    output_type : {"numpy", "sparse", "pandas"}
+        The type of the arrays that will be returned. The possibilities are:
+
+        - `"numpy"`: both `X` and `y` will be NumPy arrays;
+        - `"sparse"`: `X` will be sparse matrix and `y` will be a NumPy array;
+        - `"pandas"`: `X` will be a pandas DataFrame and `y` will be either a
+          pandas Series or DataFrame.
+
+    openml_columns_info : dict
+        The information provided by OpenML regarding the columns of the ARFF
+        file.
+
+    feature_names_to_select : list of str
+        A list of the feature names to be selected to build `X`.
+
+    target_names_to_select : list of str
+        A list of the target names to be selected to build `y`.
+
+    Returns
+    -------
+    X : {ndarray, sparse matrix, dataframe}
+        The data matrix.
+
+    y : {ndarray, dataframe, series}
+        The target.
+
+    frame : dataframe or None
+        A dataframe containing both `X` and `y`. `None` if
+        `output_array_type != "pandas"`.
+
+    categories : list of str or None
+        The names of the features that are categorical. `None` if
+        `output_array_type == "pandas"`.
+    """
+    import pandas as pd
+
+    # read the file until the data section to skip the ARFF metadata headers
+    for line in gzip_file:
+        if line.decode("utf-8").lower().startswith("@data"):
+            break
+
+    # ARFF represents missing values with "?"
+    frame = pd.read_csv(
+        gzip_file,
+        header=None,
+        na_values=["?"],  # missing values are represented by `?`
+        comment="%",  # skip line starting by `%` since they are comments
+    )
+    frame.columns = [name for name in openml_columns_info]
+
+    columns_to_select = feature_names_to_select + target_names_to_select
+    columns_to_keep = [col for col in frame.columns if col in columns_to_select]
+    frame = frame[columns_to_keep]
+
+    frame = _cast_frame(frame, openml_columns_info)
+    X, y = _post_process_frame(frame, feature_names_to_select, target_names_to_select)
+    categories = {
+        col_name: frame[col_name].cat.categories.tolist()
+        for col_name in frame.columns
+        if hasattr(frame[col_name].dtype, "is_dtype")
+        and frame[col_name].dtype.is_dtype("category")
+    }
```

Comment:
I think we can simplify this:

```python
    categories = {
        name: dtype.categories.tolist()
        for name, dtype in df.dtypes.items() if dtype == "category"
    }
```

Also can we move to a few lines below:

```python
    categories = {...}
    return X, y, None, categories
```

The categories are not needed when `output_type == "pandas"`

Reply:
With older NumPy versions it will fail:

```pytb
    categories = {
        name: dtype.categories.tolist()
        for name, dtype in frame.dtypes.items()
>       if dtype == "category"
    }
E   TypeError: data type "category" not understood

.0         = <zip object at 0x7f56713dd400>
dtype      = dtype('int64')
name       = 'legs'
```

I assume that somehow NumPy does something like `dtype("category")` to make the comparison when the type is `np.int64`.

Reply:
How about [pandas.api.types.is_categorical_dtype](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.api.types.is_categorical_dtype.html) since `pd` is already imported?

```python
    categories = {
        name: dtype.categories.tolist()
        for name, dtype in frame.dtypes.items()
        if pd.api.types.is_categorical_dtype(dtype)
    }
```